libexec/rtld-elf/rtld_c18n.h

```diff
@@ -45,113 +44,60 @@ extern const char *ld_compartment_unwind;
 /*
  * Policies
  */
+/*
+ * RTLD is the first compartment.
+ */
+#define	RTLD_COMPART_ID		0
+
 typedef uint16_t compart_id_t;
-#define	C18N_COMPARTMENT_ID_MAX	(UINT16_MAX >> 1)
+/*
+ * Define another type for the stack table index to avoid confusion.
+ */
+typedef struct { uint16_t val; } stk_table_index;
 
 compart_id_t compart_id_allocate(const char *);
 
 /*
  * Stack switching
  */
-/*
- * This macro can only be used in a function directly invoked by a trampoline.
- */
-#define	get_trusted_stk()	({					\
-		struct trusted_frame *_tf;				\
-		_Pragma("clang diagnostic push");			\
-		_Pragma("clang diagnostic ignored \"-Wframe-address\"");\
-		_tf = __builtin_frame_address(1);			\
-		_Pragma("clang diagnostic pop");			\
-		_tf;							\
-	})
+struct stk_table_stk_info {
+	size_t size;
+	void *begin;
+};
+
+struct stk_table_metadata {
+	size_t capacity;
+	struct tcb_wrapper *wrap;
+	struct stk_table_stk_info trusted;
+	struct stk_table_stk_info compart[];
```

Comment:
Please, just name them what they are and put something like stack in the name of the field. stk_table_metadata->compart is not accurate, that sounds like it's getting the compartment itself, not its stack.

Though I have to say I really do not understand why we have stacks here but also a stk_table_entry per compartment in stk_table itself. Which says to me that this needs some non-empty subset of better naming, comments and a different design.

Reply:
I've changed the names and added some comments in the struct definitions.